pkg/azuredx/azuredx.go

```diff
@@ -0,0 +1,178 @@
+type dataSourceData struct {
+	ClientID        string `json:"clientId"`
+	TenantID        string `json:"tenantId"`
+	ClusterURL      string `json:"clusterUrl"`
+	DefaultDatabase string `json:"defaultDatabase"`
+	Secret          string `json:"-"`
+}
+
+// Client is an http.Client used for API requests.
+type Client struct {
+	*http.Client
+	*dataSourceData
+	Log hclog.Logger
+}
+
+// RequestPayload is the information that makes up a Kusto query for Azure's Data Explorer API.
+type RequestPayload struct {
+	DB         string `json:"db"`
+	CSL        string `json:"csl"`
+	Properties string `json:"properties"`
+}
+
+// newDataSourceData creates a dataSourceData from the plugin API's DatasourceInfo's
+// JSONData and Encrypted JSONData which contains the information needed to connected to
+// the datasource.
+func newDataSourceData(dInfo *datasource.DatasourceInfo) (*dataSourceData, error) {
+	d := dataSourceData{}
+	err := json.Unmarshal([]byte(dInfo.GetJsonData()), &d)
+	if err != nil {
+		return nil, err
+	}
+	d.Secret = dInfo.GetDecryptedSecureJsonData()["clientSecret"]
+	return &d, nil
+}
+
+// NewClient creates a new Azure Data Explorer http client from the DatasourceInfo.
+// AAD OAuth authentication is setup for the client.
+func NewClient(ctx context.Context, dInfo *datasource.DatasourceInfo, logger hclog.Logger) (*Client, error) {
+	c := Client{}
+	c.Log = logger
+	var err error
+	c.dataSourceData, err = newDataSourceData(dInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	conf := clientcredentials.Config{
+		ClientID:     c.ClientID,
+		ClientSecret: c.Secret,
+		TokenURL:     microsoft.AzureADEndpoint(c.TenantID).TokenURL,
+		Scopes:       []string{"https://kusto.kusto.windows.net/.default"},
+	}
+
+	c.Client = conf.Client(ctx)
+	return &c, nil
+}
+
+// TestRequest handles a data source test request in Grafana's Datasource configuration UI.
+func (c *Client) TestRequest() error {
+	var buf bytes.Buffer
+	err := json.NewEncoder(&buf).Encode(RequestPayload{
+		CSL: ".show databases schema",
+		DB:  c.DefaultDatabase,
+	})
+	if err != nil {
+		return err
+	}
+
+	resp, err := c.Post(c.ClusterURL+"/v1/rest/query", "application/json", &buf)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode > 299 {
+		return fmt.Errorf("HTTP error: %v", resp.Status)
+	}
+	return nil
+}
+
+// KustoRequest executes a Kusto Query language request to Azure's Data Explorer V1 REST API
+// and returns a TableResponse. If there is a query syntax error, the error message inside
+// the API's JSON error response is returned as well (if available).
+func (c *Client) KustoRequest(payload RequestPayload) (*TableResponse, string, error) {
+	var buf bytes.Buffer
+	err := json.NewEncoder(&buf).Encode(payload)
+	if err != nil {
+		return nil, "", err
+	}
+
+	req, err := http.NewRequest(http.MethodPost, c.ClusterURL+"/v1/rest/query", &buf)
+	if err != nil {
+		return nil, "", err
+	}
+	req.Header.Set("Accept", "application/json")
```

Comment:
According to https://docs.microsoft.com/en-us/azure/kusto/api/rest/request#request-headers setting Host header is mandatory.

What about adding
- req.Header.Set("Accept-Encoding", "gzip, deflate") to get gzip'ed encoded payload (there rest of the function needs to handle that as well)
- req.Header.Set("x-ms-client-request-id", "KGC.execute;"+uuid.Must(uuid.NewV4()).String
()) to set a unique identifier for each request (exactly like the other Kusto client libraries do)

Reply:
Will open a separate issue for gzip header. Added uuid. Thanks!